### PR TITLE
Remove usage of flatRawNulls from SparkSQL functions

### PR DESCRIPTION
**Pull request by mbasmanova**
Replace usage of BaseVector::flatRawNulls() with DecodedVector::nulls().

See https://github.com/facebookincubator/velox/discussions/2202 for context.

### CODE REVIEW

File: velox/vector/BaseVector.cpp [hunk 1]
```diff
@@ -437,6 +437,28 @@ std::string BaseVector::toString() const {
     out << countNulls(nulls_, 0, length_) << " nulls";
   }
   out << "]";
+
+  return out.str();
+}
+
+std::string BaseVector::toString(bool recursive) const {
```
Comment (kgpai) [inline velox/vector/BaseVector.cpp:444]: I believe these are changes from #2206 .

File: velox/vector/DecodedVector.cpp [hunk 2]
```diff
@@ -514,4 +516,35 @@ VectorPtr DecodedVector::wrap(
...
+
+  if (hasExtraNulls_) {
+    allNulls_ = nulls_;
+  } else if (!nulls_ || size_ == 0) {
+    allNulls_ = nullptr;
+  } else {
+    if (isIdentityMapping_) {
+      allNulls_ = nulls_;
+    } else if (isConstantMapping_) {
+      copiedNulls_.resize(0);
```
Comment (kgpai) [inline velox/vector/DecodedVector.cpp:533]: why resize to 0 and then resize to nwords(size_) ? Why not do that in first place ?

File: velox/vector/DecodedVector.cpp [hunk 3]
```diff
@@ -514,4 +516,35 @@ VectorPtr DecodedVector::wrap(
...
+      copiedNulls_.resize(bits::nwords(size_));
+      auto* rawCopiedNulls = copiedNulls_.data();
+      for (auto i = 0; i < size_; ++i) {
+        bits::setNull(rawCopiedNulls, i, bits::isBitNull(nulls_, indices_[i]));
+      }
+      allNulls_ = copiedNulls_.data();
+    }
+  }
+
+  allNullsInitialized_ = true;
```
Comment (kgpai) [inline velox/vector/DecodedVector.cpp:547]: Maybe allNulls_ can be an optional so that we dont have to use allNullsInitialzed_ can just check if allNulls_.has_value()

File: velox/vector/DecodedVector.h [hunk 4]
```diff
@@ -283,10 +265,21 @@ class DecodedVector {
...
+  // (hasExtraNulls_ is false). Otherwise, null bitmask of the base vector
+  // combined with null bitmasks in all the wrappings (hasExtraNulls_ is true).
+  // May be nullptr if there are no nulls.
   const uint64_t* nulls_ = nullptr;
 
+  // Nulls bitmask indexed using top-level row numbers and containing null bits
+  // of the base vector combined with null bits in all the wrappings. May be
+  // nullptr if there are no nulls or allNullsInitialized_ is false. Initialized
+  // on first access.
+  const uint64_t* allNulls_ = nullptr;
```
Comment (kgpai) [inline velox/vector/DecodedVector.h:278]: Maybe have this as an optional so we wont need allNullsinitialized_.

File: velox/vector/DecodedVector.cpp [hunk 5]
```diff
@@ -514,4 +516,35 @@ VectorPtr DecodedVector::wrap(
       rows.end(),
       std::move(data));
 }
+
+const uint64_t* DecodedVector::nulls() {
```
Comment (kgpai) [inline velox/vector/DecodedVector.cpp:520]: These changes probably belong in a different PR ?

File: velox/vector/BaseVector.cpp (hunk 1 above)
Comment (mbasmanova) [reply velox/vector/BaseVector.cpp:444]: This PR depends on 2 other PRs: #2206 and #2205. The last commit contains the actual changes.

File: velox/vector/DecodedVector.cpp (hunk 2 above)
Comment (mbasmanova) [reply velox/vector/DecodedVector.cpp:533]: To make sure that all values are set to bits::kNull64.

File: velox/vector/DecodedVector.cpp (hunk 3 above)
Comment (mbasmanova) [reply velox/vector/DecodedVector.cpp:547]: That's a good idea.

File: velox/vector/DecodedVector.cpp (hunk 5 above)
Comment (mbasmanova) [reply velox/vector/DecodedVector.cpp:520]: Indeed. See my reply above.

**Comment by mbasmanova** [comment]
@kgpai Thank you for review. I addressed the comments in #2205

**Review by xiaoxmeng — APPROVED** [review]

File: velox/vector/BaseVector.cpp (hunk 1 above)
Comment (kgpai) [reply velox/vector/BaseVector.cpp:444]: Thanks Masha, will wait for those 2 prs to be merged.
